fix(salt): check bash version is modern, esp macos

noelmcloughlin opened the pull request:
This PR checks assumption _a modern `bash version 4 or later` is installed_ and addresses `line 23: declare: -A: invalid option` errors with old bash.
 
For example, a newly installed MacOS has old bash version (see: https://itnext.io/upgrading-bash-on-macos-7138bd1066ba).

PR verified on Ubuntu/MacOS.

wisererik commented on salt/salter.sh:
```diff
@@ -190,7 +197,7 @@ salt-bootstrap() {
              sudo python ${PWD}/get-pip.py 2>/dev/null
 
-             [ -x /usr/local/bin/brew ] | su - ${USER} -c '/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"'
+             # [ -x /usr/local/bin/brew ] | su - ${USER} -c '/usr/bin/ruby -e "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/install)"'
```
please remove unused code here :)

noelmcloughlin replied:
Done, thanks for spotting that @wisererik

noelmcloughlin commented:
Added a commit to fix a bug https://github.com/saltstack-formulas/salter/issues/31
- Fix bug causing *unnecessary recloning* of opensds-formula duirng opensds install
- merge minor upstream changes

noelmcloughlin commented:
Verified on CentOS
```
[vagrant@centos7 ~]$ sudo bash salter.sh -i opensds
cloning opensds-installer from opensds ...

local:
    ----------
    base:
        - opensds.infra
run salt: this takes a while, please be patient ...
--------------
Succeeded: 106 (unchanged=3, changed=56)
Failed:      0
--------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201812 ]

/////////////////////////////////////////////////////////////////
        infra for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.telemetry
run salt: this takes a while, please be patient ...
-------------
Succeeded: 41 (changed=35)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201928 ]

/////////////////////////////////////////////////////////////////
        telemetry for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.keystone
run salt: this takes a while, please be patient ...
------------
Succeeded: 1
Failed:    0
------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201931 ]

/////////////////////////////////////////////////////////////////
        keystone for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.config
run salt: this takes a while, please be patient ...
-------------
Succeeded: 15 (changed=9)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201932 ]

/////////////////////////////////////////////////////////////////
        config for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.database
run salt: this takes a while, please be patient ...
-------------
Succeeded: 34 (unchanged=2, changed=15)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201932 ]

/////////////////////////////////////////////////////////////////
        database for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.auth
run salt: this takes a while, please be patient ...
-------------
Succeeded: 27 (changed=12)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201933 ]

/////////////////////////////////////////////////////////////////
        auth for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.hotpot
run salt: this takes a while, please be patient ...
-------------
Succeeded: 30 (changed=14)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201933 ]

/////////////////////////////////////////////////////////////////
        hotpot for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.sushi
run salt: this takes a while, please be patient ...
-------------
Succeeded: 57 (changed=30)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201935 ]

/////////////////////////////////////////////////////////////////
        sushi for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.backend
run salt: this takes a while, please be patient ...
-------------
Succeeded: 85 (changed=58)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908201938 ]

/////////////////////////////////////////////////////////////////
        backend for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.dock
run salt: this takes a while, please be patient ...
-------------
Succeeded: 30 (changed=14)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908202001 ]

/////////////////////////////////////////////////////////////////
        dock for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.dashboard
run salt: this takes a while, please be patient ...
-------------
Succeeded: 21 (changed=5)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908202001 ]

/////////////////////////////////////////////////////////////////
        dashboard for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.gelato
run salt: this takes a while, please be patient ...
-------------
Succeeded: 30 (changed=13)
Failed:     0
-------------
See full log in [ /tmp/opensds-opensds-installer/log.201908202002 ]

/////////////////////////////////////////////////////////////////
        gelato for opensds-installer has completed
////////////////////////////////////////////////////////////////

local:
    ----------
    base:
        - opensds.freespace
run salt: this takes a while, please be patient ...
------------
Succeeded: 4 (changed=1)
Failed:    0
------------
See full log in [ /tmp/opensds-opensds-installer/log.201908202006 ]

/////////////////////////////////////////////////////////////////
        freespace for opensds-installer has completed
////////////////////////////////////////////////////////////////
```
